### DISCUSSION TITLE
[Idea] Restrict debug mode to an ip address or admin account

### DISCUSSION THREAD

**Discussion by jf-viguier** [PrestaShop/PrestaShop#30840] [Ideas]
Debug mode (and profiling mode) is sometime useful in production but it may cause some error display in front (ans security flow).

IMO, the debug mode should be restricted to :
1) an ip address or address range (ip v6, whole company)
2) An admin account : better way but harder to code.
3) both

Detect and admin account in front is not easy : I was thinking of a token stored in a cookie.

Debug mode is set before SF load, so it has to be in pure php.

WDYT ?

**Comment by StoreCommanderDev** [comment]
Yes this is a good idea.

How many times we need to do a condition on our IP or with a specific cookie

**Comment by FabienPapet** [comment]
👋 For what purpose do you need it (even sometimes) ? Smarty debugging ? Other ?

**Reply by jf-viguier** [reply]
smarty debugging, module debugging, profiling to detect bottleneck,  ...

**Comment by studiokiwik** [comment]
It would be very useful. Many problems only occur in the production environment.

**Comment by hrostyk** [comment]
In a situation where you need to quickly find an error, this function can be useful.

This function can be implemented:
1. Use the IP address from the Maintenance page.
2. To prevent Debug Mode from being shown for employees, add permissions for Debug Mode in to Profiles.

**Reply by Hlavtox** [reply]
You can just forbid the from the whole performance page, they shouldn't touch it anyway.

**Comment by unlocomqx** [comment]
I've been thinking about this since I need to enable debug mode only for myself when debugging a customer's shop. 
I use xDebug Helper extension which sets a cookie, something like XDEBUG_SESSION. 
I have to modify the defines file manually, so a backoffice option would be convenient for me

<img width="488" alt="CleanShot 2023-02-10 at 11 20 21@2x" src="https://user-images.githubusercontent.com/793712/218067230-e9a21970-b1c4-4eb4-b06a-c79bc4832126.png">

```php
define('_PS_MODE_DEV_', isset($_COOKIE['XDEBUG_SESSION']));
```

> Detect and admin account in front is not easy

It's quite [easy](https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/front/UploadController.php#L53) actually 

So I guess, these options would be useful
- Pick a profile (superadmin for example)
- Enter IP
- Enter cookie name that should be present

and then the system would check if any criteria is met and thus enable debug mode

**Reply by PrestaEdit** [reply]
Now that the maintenance mode is available without write it's IP, when you are a logged employee, the detection system maybe could be re-used for this.

**Comment by MatShir** [comment]
I see great ideas in these threads. How can we make this idea become a cool feature for the v9. Any members willing to contribute ?

**Comment by unlocomqx** [comment]
@MatShir I will give it a try then we can discuss the specifics

Cheers

**Comment by unlocomqx** [comment]
I submitted PR https://github.com/PrestaShop/PrestaShop/pull/32809

**Comment by lmeyer1** [comment]
I see a problem with debug mode:

In the BO, when an erro happens, a nice page displays, that offers the option to activate debug mode.

This is fallacious, because a merchant may turn on debug mode without beeing aware of what the consequences are for the installation. It is not obvious to detect that debug mode is on, but it can have adverse effects on front office.

**Reply by MatShir** [reply]
True that sometime merchant don't realize the consequences. Sometime, dev can ask for the debug mode to see the error message. Merchant or dev can forget to deactivate it. 
We could probably add more awareness and warning like: visual layout on the BO when such feature is on, confirmation modal or admin access only.

**Comment by jf-viguier** [comment]
PR #32809 is a first step on this topic and is included in Version 9+, thanks @unlocomqx
